Wikipedia/Code/TalkPageContainerViewController.swift

```diff
@@ -191,6 +192,121 @@ private extension TalkPageContainerViewController {
+                self.pushTalkPage(title: titleWithTalkPageNamespace, siteURL: siteURL)
+            }
+        }
+        let cancelAction = UIAlertAction(title: CommonStrings.cancelActionTitle, style: .cancel, handler: nil)
+        
+        alertController.addAction(safariAction)
+        alertController.addAction(talkAction)
+        alertController.addAction(cancelAction)
+        
+        present(alertController, animated: true, completion: nil)
+    }
+    
+    func openURLInSafari(url: URL) {
+        UIApplication.shared.open(url, options: [:], completionHandler: nil)
+    }
+    
+    func toggleLinkDeterminationState(loadingViewController: FakeLoading & ViewController, shouldDisable: Bool) {
+        
+        if shouldDisable {
+            loadingViewController.fakeProgressController.start()
+        } else {
+            loadingViewController.fakeProgressController.stop()
+        }
+        
+        loadingViewController.scrollView?.isUserInteractionEnabled = !shouldDisable
+        loadingViewController.navigationItem.rightBarButtonItem?.isEnabled = !shouldDisable
+    }
+    
+    func tappedLink(_ url: URL, loadingViewController: FakeLoading & ViewController) {
+        guard let absoluteURL = absoluteURL(for: url) else {
+            showNoInternetConnectionAlertOrOtherWarning(from: TalkPageError.unableToDetermineAbsoluteURL)
+            return
+        }
+        
+        toggleLinkDeterminationState(loadingViewController: loadingViewController, shouldDisable: true)
+        
+        self.dataStore.articleSummaryController.updateOrCreateArticleSummariesForArticles(withURLs: [absoluteURL]) { [weak self, weak loadingViewController] (articles, error) in
+            
+            guard let self = self,
+            let loadingViewController = loadingViewController else {
+                return
+            }
+            
+            self.toggleLinkDeterminationState(loadingViewController: loadingViewController, shouldDisable: false)
+            
+            if error != nil {
+                self.openURLInSafari(url: absoluteURL)
+                return
+            }
+            
+            if let namespace = articles.first?.pageNamespace,
+                
+                //convert absoluteURL into a siteURL that TalkPageType.user.titleWithCanonicalNamespacePrefix will recognize to prefix the canonical name
+                let languageCode = absoluteURL.wmf_language,
+                let siteURL = MWKLanguageLinkController.sharedInstance().language(forLanguageCode: languageCode)?.siteURL() {
+                
+                switch namespace {
+                case .userTalk:
+                    let lastPathComponent = url.lastPathComponent
+                    self.pushTalkPage(title: lastPathComponent, siteURL: siteURL)
+                case .user:
+                    self.showUserActionSheet(siteURL: siteURL, absoluteURL: absoluteURL)
+                default:
```

Comment:
Should there be a case above `default` for `.main` to show an article in an article view controller instead of bumping to Safari?

Reply:
@joewalsh agreed!